[RFC] <#239> Add Contributor Calendar

federicobucchi opened the pull request:
### Motivation:

This PR has been made in relation to this issue: https://github.com/apple/swift-org-website/issues/239

Please leave feedbacks about the scaffold and we will push new changes.

Note:
- Design of the calendar is not final (there are going to be color adjustments)
- Currently the events in the calendar are placeholders.. we are working on adding real events soon
- More event fields will be displayed in the RSS feed, in the calendar event details UI and in the iCal data (it will be done when we have all the first batch of real events to add)
- The web components repository is still private, it will be opened in the near future (but after the inclusion of this Calendar on the website)

The following PR is:
- Adding a new menu link called "Calendar"
- Adding a new route`/calendar`
- Include a web components library with currently only 2 web components (`<swift-calendar />` and `<swift-modal />`) that eventually will be shared for contributions (`<swift-modal />` is used inside `<swift-calendar />` for event details)
- - Showing a calendar like view with days/months, a list of events by month view and an event details view
- Supporting light and dark mode
- Supporting different viewports
- Allowing  owners to change their calendar via a YAML file
- Allowing viewers to download an iCal as a subscription so viewers can subscribe to the entire calendar downloading an .ics file and deciding how often to pull for updates
- Adding a new link (RSS icon) to allow users to subscribe to changes to the calendar (the link/icon is not visible in the screenshots below just because the screenshots have not been updated after this addition)

### Modifications:

- Updated README.md to indicate what YAML file users can modify to edit/add/remove events to the calendar
- Updated existing files to add new menu item and new route `/calendar`
- Added `_data/calendar.yml` (file that users can modify to edit/add/remove events to the calendar)
- Added files and assets of the web components in a new folder `assets/javascripts/swift-web-components/`
- Added Jekyll generator for SwiftEvents.ics (generated .ics for viewers to download as subscription)

### Result:

## Desktop (Light Theme)
![7934da8c-3399-461c-b011-ed4cfa3289f1](https://user-images.githubusercontent.com/59409/235494260-8357447b-3223-47a6-bd87-a798f381a48b.png)

![1b4dd2e0-7d8d-4931-91a8-1f8c3afbf06e](https://user-images.githubusercontent.com/59409/235494300-3b8ca3bd-cbe3-4ba3-b4a9-cd338aefc4b2.png)

![49296980-7577-4e4a-a78d-e8f16f4a3507](https://user-images.githubusercontent.com/59409/235494347-990f9070-9ded-4e83-b8e0-a99d1abb4bcf.png)

## Desktop (Dark Theme)
![2f350fd6-1256-4e6d-a822-1aceffd44955](https://user-images.githubusercontent.com/59409/235494554-e87bfe6e-7bce-4dcd-9a45-49858b88e50f.png)

![9f3946e5-6787-49f2-bbec-e61d6370963b](https://user-images.githubusercontent.com/59409/235494589-1b064f75-0bda-4158-9636-d42a9a8f022a.png)

![31212a72-99f2-43f9-b016-710583fae91a](https://user-images.githubusercontent.com/59409/235494722-9bbae053-98f3-4acb-a754-604a3faa3046.png)

## Mobile (Light Theme)
![afaa7099-963c-4fc3-9776-734ecd8573af](https://user-images.githubusercontent.com/59409/235495229-9a0f27cf-cfc7-408d-9ec7-c20b4a679d9f.png)

![824ac610-21fd-4d4e-b6ef-3d6ca64689a2](https://user-images.githubusercontent.com/59409/235495262-638f99be-f974-4d7a-9535-324a0917648e.png)

![05dd2a4a-b4d9-43e9-8259-d161aaa297db](https://user-images.githubusercontent.com/59409/235495283-f532216e-88bc-40d5-a706-e835b80b15d9.png)

## Mobile (Dark Theme)
![339177af-4d5b-47dd-8a3f-d5196d8d4b91](https://user-images.githubusercontent.com/59409/235495341-9cf889df-175f-44f0-a932-604da7d5c234.png)

![43faa1d4-3e93-4a56-9525-aa5e01ad6d8b](https://user-images.githubusercontent.com/59409/235495420-773b3ebf-dd0d-4465-afa4-c5b22ee6725a.png)

![d1ecf202-c577-41db-9361-8c039d0da49b](https://user-images.githubusercontent.com/59409/235495449-00e23f1f-c79b-4806-b8ef-332a827045db.png)

dempseyatgithub commented:
First, thank you for the work you've put into this. I think a calendar will be very beneficial for the community.

> * Include a web components library with currently only 2 web components (`<swift-calendar />` and `<swift-modal />`) that eventually will be shared for contributions (`<swift-modal />` is used inside `<swift-calendar />` for event details)

Could you give us an overview of how you are envisioning the web components library for the site to work overall?

Is the intent for the web components to be a separate public repository or would they be added to the swift-org-website repository?

> The web components repository is still private, it will be opened in the near future (but after the inclusion of this Calendar on the website)

Is there a reason the implementation of the web components will be private until after we've already added them to the site? It seems like that should be part of the PR review for this addition.

> Include a web components library with currently only 2 web components (`<swift-calendar />` and `<swift-modal />`)

One other point is that with the names `swift-calendar` and `swift-modal` I initially thought that possibly the components were somehow written in Swift.

Once the Swift.org infrastructure moves to something implemented in Swift, this might be more of a point of confusion.

It might be more clear to use `swift-org` as the prefix: `swift-org-calendar`, `swift-org-modal`.

yakatyansh commented:
@federicobucchi  I see you have pushed a calendar, any help or addition I can do to this  ?

alexandersandberg reviewed (commented):
This is a great start, @federicobucchi! Well done.

I noticed that the events modal needs some height adjustments when opened with the calendar with the events view enabled.

This seems to be due to the modal not being able to extend outside the calendar component, and the calendar component being very small when only displaying events (as setting an explicit height on the calendar component fixes this).

<img width="592" alt="CleanShot 2023-05-02 at 11 13 22" src="https://user-images.githubusercontent.com/35671299/235627651-b40979fc-fa2d-4c7e-b44b-ac590f1afe6b.png">

federicobucchi commented:
> Could you give us an overview of how you are envisioning the web components library for the site to work overall?

The component library is going to be (eventually) a standalone mono repo that will have components developed in Stencil / TypeScript but exported for being included directly on "native" web page projects or frameworks like React, Vue, Angular (trying to be as agnostic as possible).

> Is the intent for the web components to be a separate public repository or would they be added to the swift-org-website repository?

Correct, the web components will be (eventually) a separate repository for development and the components will be eventually importable in different ways: via a `script` tag or an NPM package.  For this first phase, I have only exported the distribution files and included them directly in the website (this is why they are just all minified).  With the current Ruby / Jekyll app it seems the approach to go, but, even if we redesign the website to use another framework or we decide to write the website in Swift, these web components can still be imported and used in the new project.

> Is there a reason the implementation of the web components will be private until after we've already added them to the site? It seems like that should be part of the PR review for this addition.

I will update you about this later on, gathering info and things

> One other point is that with the names `swift-calendar` and `swift-modal` I initially thought that possibly the components were somehow written in Swift.
> Once the Swift.org infrastructure moves to something implemented in Swift, this might be more of a point of confusion.
> It might be more clear to use `swift-org` as the prefix: `swift-org-calendar`, `swift-org-modal`.

True, naming can be changed.  I didn't want to have something to broad like `web-components` but also I didn't want to be too specific `swift-org` because I wanted these to potentially be used around more swift websites/resources etc (so not really for any website but also not only for [swift.org](http://swift.org/))

I would focus on gathering feedbacks specifically about the calendar, later on, reviving the topic of web components mono repo.

federicobucchi commented:
@yakatyansh

> @federicobucchi I see you have pushed a calendar, any help or addition I can do to this ?
I hope you remember me

Thank you!

I am gathering feedbacks to have an idea of where help/additions are needed! I will circle back later on!

tomerd commented on _data/navigation.yml:
```diff
@@ -8,6 +8,8 @@
           name: swiftorg-and-open-source
         - title: Platform Support
           name: platform-support
+    - title: Calendar
```
this feels like the wrong location to me, perhaps it should be under community as a sub item

federicobucchi replied:
Sure, can you propose exactly in between which items?

tomerd commented on _layouts/base.html:
```diff
@@ -74,6 +74,8 @@
 {{content}}
 <script src="/assets/javascripts/application.js"></script>
+<script type="module" src="/assets/javascripts/swift-web-components/swift-web-components-stencil.esm.js"></script>
+<script nomodule src="/assets/javascripts/swift-web-components/swift-web-components-stencil.js"></script>
```
can this need be done in a more specific place for the calendar instead of main base layout?

federicobucchi replied:
I can try to add it here: calendar/index.html

federicobucchi replied:
> another comment, all of this javascript and css should be put under a "calendar" directory and not a "swift-web-components" director, the latter is too broad / generic of a name

This is including calendar and modal components, so not only calendar

tomerd replied:
we should limit the scope here to calendar

federicobucchi replied:
I am using the modal component I made to open the details of the calendar event

tomerd replied:
the point I am making is these files are all related to calendar and should be in a directory named calendar and not "swift-web-components"

yakatyansh replied:
> the point I am making is these files are all related to calendar and should be in a directory named calendar and not "swift-web-components", the files also should be named calendar and not "swift-web-components". iow the _naming_ should be focused on calendar and not sound generic


I would like to suggest a naming scheme, can we name calendar-components because it not only consists a calendar but also supporting components like modal, so calendar-components is good in my opinion

tomerd commented on assets/javascripts/swift-web-components/components/files/feed.xml:
```diff
@@ -0,0 +1,22 @@
+    <atom:link href="{{ site.url }}/assets/javascripts/swift-web-components/components/files/feed.xml" rel="self" type="application/rss+xml" />
+    <pubDate>{{ site.time | date_to_rfc822 }}</pubDate>
+    <lastBuildDate>{{ site.time | date_to_rfc822 }}</lastBuildDate>
+    {% for calendar in site.data.calendar %}
```
doe jekyll generate a feed from this file?

federicobucchi replied:
Correct

federicobucchi commented on calendar/index.html:
```diff
@@ -0,0 +1,19 @@
+
+<ul class="calendar-menu">
+  <li>
+    <a href="/assets/javascripts/swift-web-components/components/files/feed.xml" title="Subscribe to Calendar Updates">
```
With a Jekyll generator: assets/javascripts/swift-web-components/components/files/feed.xml

yakatyansh commented:
> @yakatyansh
> 
> > @federicobucchi I see you have pushed a calendar, any help or addition I can do to this ?
> > I hope you remember me
> 
> Thank you!
> 
> I am gathering feedbacks to have an idea of where help/additions are needed! I will circle back later on!

Sure, Do tell me wherever I can contribute to this issue, would love to do so,

yakatyansh commented:
> @alexandersandberg
> 
> Thanks for trying it!
> 
> You are correct, I am putting this on the todo list!

I would like to work on that as a starter

yakatyansh commented:
@federicobucchi I was trying to pull this branch on my machine but was unable to, how should I do so ?
pls guide me 

I have tried git pull

federicobucchi commented:
> @federicobucchi I was trying to pull this branch on my machine but was unable to, how should I do so ? pls guide me
> 
> I have tried git pull

Hello @yakatyansh, if you don't have experience with Git, please checkout some tutorials: https://www.w3schools.com/git/

Maybe you have to fetch first:
```
git fetch origin

or

git fetch origin
```

then

```
git checkout fb-include-calendar
```

For Git support, you are probably have more success asking to the community that only me.

Cheers!

yakatyansh commented:
Was able to fetch it, thank you so much 👍

yakatyansh commented:
https://github.com/apple/swift-org-website/assets/90173866/12cf2da7-914e-4538-ab06-e9245c524a1a

is it on my device or this is a bug I wonder ?

federicobucchi commented:
@yakatyansh 

> Screen.Recording.2023-05-10.at.7.15.39.PM.mov 
> 
> is it on my device or this is a bug I wonder ?

I tried with different browser and I did experience that.  It seems that the component is not able to understand what "Appearance" you are on (Light, Dark, Auto).  Do you have any browser extension that might be playing with the theme?

federicobucchi commented:
@alexandersandberg I pushed a new version and it shouldn't have the issue you described above!  Please let me know if you are able to verify! Thank you!

@tomerd I made some changes, and I have:
- Moved the things `scripts` inclusion inside the specific file `calendar/index.html`
- I renamed `swift-web-components` to `web-components` (in case we keep this framework or we go to a new one, the usage and development of more components will not be affected and this folder and naming will allow us to extend including more components without having to rename again).  Just calling it `calendar` or `calendar-component` currently doesn't embrace the fact that the modal component can be used standalone, which it can)
- Currently I have moved the view under Community but not inside Community Overview, but between Community Overview and Diversity.  The reason why I didn't, is that it seems the calendar to get a bit lost in between subsections of a section, the page seems to start becoming very long with a lot of scroll and subsections and the file is a `markdown` (which I generally don't like to mix with HTML if possible):

<img width="1080" alt="Screenshot 2023-05-12 at 4 20 24 PM" src="https://github.com/apple/swift-org-website/assets/59409/c1bce416-2d4f-4a32-a26d-59c7780f0961">

<img width="1043" alt="Screenshot 2023-05-12 at 4 20 36 PM" src="https://github.com/apple/swift-org-website/assets/59409/c8aa98d4-5c19-43e9-a345-7425fb5a657a">

yakatyansh commented:
> @yakatyansh
> 
> > Screen.Recording.2023-05-10.at.7.15.39.PM.mov
> > is it on my device or this is a bug I wonder ?
> 
> I tried with different browser and I did experience that. It seems that the component is not able to understand what "Appearance" you are on (Light, Dark, Auto). Do you have any browser extension that might be playing with the theme?

nope, I was using brave browser without any extensions

federicobucchi commented:
@yakatyansh I verified the issue on Brave browser.  Can you please pull and try to run it again?

alexandersandberg commented:
>@alexandersandberg I pushed a new version and it shouldn't have the issue you described above! Please let me know if you are able to verify! Thank you!

The issue is fixed—thanks!

yakatyansh reviewed (commented):
https://github.com/apple/swift-org-website/assets/90173866/598e3c88-b05b-44bd-b1cb-76ab09775819

So, this is my observation, after changing themes everything turns to normal but if once again  refresh the page, this is what happens

federicobucchi commented:
> Screen.Recording.2023-05-16.at.1.04.32.PM.mov 
> 
> So, this is my observation, after changing themes everything turns to normal but if once again refresh the page, this is what happens

After my latest changes, also Brave browser is fixed for me.

Can you pull the changes:

```
git pull origin fb-include-calendar
```

then (if you are using Docker rebuild and re-run

```
docker-compose run build
```

then

```
docker-compose up website
```

yakatyansh commented:
Oh my bad, I had some merge conflicts on my local, fixed them, commit them but didn't push, if you allow me I will push them

yakatyansh commented:
the brave browser issue is fixed now, actually I want to try some XML, if you could delegate me some tasks of including the events as I can see now the events are just placeholders in the calendar, it would be great

federicobucchi commented:
> Oh my bad, I had some merge conflicts on my local, fixed them, commit them but didn't push, if you allow me I will push them

@yakatyansh you should't have any conflict, what changes do you need to push?
I wouldn't add functionalities right now, but only fixes if you see something wrong

> the brave browser issue is fixed now, actually I want to try some XML, if you could delegate me some tasks of including the events as I can see now the events are just placeholders in the calendar, it would be great

`XML` there is not much right now until we get the initial events to add.  I don't have them yet.
The calendar might potentially go live with no events and then have people adding them

yakatyansh commented:
> @yakatyansh you should't have any conflict, what changes do you need to push? I wouldn't add functionalities right now, but only fixes if you see something wrong
> 

initially I was unable to pull I got this error 
<img width="513" alt="image" src="https://github.com/apple/swift-org-website/assets/90173866/4d19bfb6-9b2d-4358-8832-44ec991a6389">
I fixed that using ``git config --global  pull.ff true``
then I pulled it again 
<img width="688" alt="image" src="https://github.com/apple/swift-org-website/assets/90173866/ca219bc8-24f7-4e3c-a487-e31f150cc082">

so I fixed this merge conflict along with some xml files 
<img width="618" alt="image" src="https://github.com/apple/swift-org-website/assets/90173866/308d75e9-fc7e-4428-bd50-862418029522">

this is what I committed. after that all was good on my brave browser

federicobucchi commented:
@yakatyansh I was confused when you said `XML` instead of `YAML`.

I wouldn't push anything seeing the files you modified:
- All those .yml files have nothing to do with this PR
- You shouldn't have conflicts

yakatyansh commented:
> @yakatyansh I was confused when you said `XML` instead of `YAML`.
> 
> 
> 
> I wouldn't push anything seeing the files you modified:
> 
> - All those .yml files have nothing to do with this PR
> 
> - You shouldn't have conflicts

Alright

shahmishal commented:
Once we have a process in place to include events, we should revisit this PR.
